### PR TITLE
[feature] drop lake table

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 4 / 57 (07.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDelete.java|2|48|04.17%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|2|50.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+public class ShardDelete extends MasterDaemon implements Writable {
+    private static final Logger LOG = LogManager.getLogger(ShardDelete.class);
+
+    @SerializedName(value = "shardIdToTablet")
+    private Map<Long, LakeTablet> shardIdToTablet;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java:39]: shard id is enough.

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    public void addShardId(long shardId, LakeTablet tablet) {
+        shardIdToTablet.put(shardId, tablet);
+    }
+
+    @Override
+    protected void runAfterCatalogReady() {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java:46]: drop tablet first, then drop shard.

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 3]
```diff
@@ -244,6 +244,15 @@ public long getWorkerIdByBackendId(long backendId) {
         }
     }
 
+    public void deleteShards(Set<Long> shardIds) throws DdlException {
+        // prepare();
+        try {
+            client.deleteShard(serviceId, shardIds);
+        } catch (StarClientException e) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:251]: maybe already not exist.

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        // delete shard and drop lakeTablet
+        Iterator<Map.Entry<Long, LakeTablet>> iterator = shardIdToTablet.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<Long, LakeTablet> entry = iterator.next();
+            // 1. delete shard
+            long shardId = entry.getKey();
+            try {
+                Set<Long> shardIds = new HashSet<>();
+                shardIds.add(shardId);
+                GlobalStateMgr.getCurrentState().getStarOSAgent().deleteShards(shardIds);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java:56]: better to batch drop the shards that associated tablets have been dropped successfully.

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                    if (!response.failedTablets.isEmpty()) {
+                        finished = false;
+                    }
+                } catch (ExecutionException | InterruptedException e) {
+                    LOG.error(e);
+                    finished = false;
+                }
+
+            } catch (UserException e) {
+                LOG.warn("failed to get primary backendId");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java:90]: finished = false;

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                address.setHostname(backend.getHost());
+                address.setPort(backend.getBrpcPort());
+                LakeService lakeService = BrpcProxy.getInstance().getLakeService(address);
+
+                DropTabletRequest request = new DropTabletRequest();
+                List<Long> tabletIds = Lists.newArrayList();;
+                tabletIds.add(shardId);
+                request.tabletIds = tabletIds;
+
+                Future<DropTabletResponse> responseFuture = lakeService.dropTablet(request);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java:78]: better to batch drop tablets.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 18 / 78 (23.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDelete.java|8|53|15.09%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|3|8|37.50%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|4|50.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 31 / 102 (30.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDelete.java|14|59|23.73%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|7|28.57%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|3|8|37.50%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|5|40.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|4|50.00%|
|:large_blue_circle:|com/starrocks/server/StarosInfo.java|5|9|55.56%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 44 / 156 (28.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|0|16|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDelete.java|15|75|20.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|3|8|37.50%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|6|12|50.00%|
|:large_blue_circle:|com/starrocks/server/StarosInfo.java|5|9|55.56%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|9|77.78%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 44 / 156 (28.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|0|16|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDelete.java|15|75|20.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|3|8|37.50%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|6|12|50.00%|
|:large_blue_circle:|com/starrocks/server/StarosInfo.java|5|9|55.56%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|9|77.78%|

File: fe/fe-core/src/main/java/com/starrocks/server/StarosInfo.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+public class StarosInfo {
+
+    private static final Logger LOG = LogManager.getLogger(NodeMgr.class);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/StarosInfo.java:15]: Not `NodeMgr.class`

File: fe/fe-core/src/main/java/com/starrocks/server/StarosInfo.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    public StarosInfo() {
+        this.shardDelete = new ShardDelete();
+    }
+
+    public ShardDelete getShardDelete() {
+        return shardDelete;
+    }
+
+    public long loadShardDeleteInfo(DataInputStream in, long checksum) throws IOException {
+        shardDelete.read(in);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/StarosInfo.java:28]: Use json to serialize this class, not only `ShardDelete`.

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+                    LOG.error(e);
+                    finished = false;
+                }
+            }
+
+            // 2. delete shard
+            try {
+                // get shard first
+                List<Long> tabletIds = new ArrayList<>(shards);
+                GlobalStateMgr.getCurrentState().getStarOSAgent().getShards(tabletIds);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java:115]: `getShards` is useless.

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                GlobalStateMgr.getCurrentState().getStarOSAgent().deleteShards(shards);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDelete.java:116]: If some tablets are dropped failed, the shards cannot be dropped.

**Pull request closed without merging**